### ISSUE TITLE
OfType<T> throws InvalidOperationException when Derived entity has Owned entity type

### ISSUE THREAD

**Issue by uso800** [aspnet/EntityFrameworkCore#13430]
When derived entity types has owned entity types, query with OfType&lt;T&gt; fails.
Query against base type work correctly.

Below is the exception message.
```
Exception message:
System.InvalidOperationException: 'The property 'OwnedEntity2' on entity type 'DerivedEntity' could not be found. Ensure that the property exists and has been included in the model.'

Stack trace:
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.MemberAccessBindingExpressionVisitor.GetPropertyPath(Expression expression, QueryCompilationContext queryCompilationContext, QuerySourceReferenceExpression& querySourceReferenceExpression)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.BindPropertyExpressionCore[TResult](Expression propertyExpression, IQuerySource querySource, Func`3 propertyBinder)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.BindNavigationPathPropertyExpression[TResult](Expression propertyExpression, Func`3 propertyBinder)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitAndConvert[T](ReadOnlyCollection`1 nodes, String callerName)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.VisitNewArray(NewArrayExpression node)
   at System.Linq.Expressions.NewArrayExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitAndConvert[T](ReadOnlyCollection`1 nodes, String callerName)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Remotion.Linq.Clauses.SelectClause.TransformExpressions(Func`2 transformation)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.NavigationRewritingQueryModelVisitor.VisitSelectClause(SelectClause selectClause, QueryModel queryModel)
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.Rewrite(QueryModel queryModel, QueryModel parentQueryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.OptimizeQueryModel(QueryModel queryModel, Boolean asyncQuery)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.OptimizeQueryModel(QueryModel queryModel, Boolean asyncQuery)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateQueryExecutor[TResult](QueryModel queryModel)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](Expression query, IQueryModelGenerator queryModelGenerator, IDatabase database, IDiagnosticsLogger`1 logger, Type contextType)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass13_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Remotion.Linq.QueryableBase`1.GetEnumerator()
   at System.Collections.Generic.List`1.AddEnumerable(IEnumerable`1 enumerable)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at ConsoleApp1.Program.Main(String[] args) in E:\src\ConsoleApp1\ConsoleApp1\Program.cs:line 18
```

### Steps to reproduce
You can reproduce this problem with the following code.

```c#
using System.Linq;
using Microsoft.EntityFrameworkCore;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            var context = new MyDbContext();
            context.Database.EnsureCreated();

            context.Add(new DerivedEntity());
            context.Add(new DerivedEntity2());
            context.SaveChanges();

            var entities = context.Entities.ToList(); // it works fine. returns 2 instances.
            var derivedEntities = context.Entities.OfType<DerivedEntity>().ToList(); // it throws exception.
        }
    }

    class MyDbContext : DbContext
    {
        public DbSet<BaseEntity> Entities { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer("Server=(localdb)\\mssqllocaldb;Database=ConsoleApp1Database;Trusted_Connection=True;MultipleActiveResultSets=true");
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<DerivedEntity>(builder =>
            {
                builder.HasBaseType<BaseEntity>();
                builder.OwnsOne(_ => _.OwnedEntity);
            });

            modelBuilder.Entity<DerivedEntity2>(builder =>
            {
                builder.HasBaseType<BaseEntity>();
                builder.OwnsOne(_ => _.OwnedEntity2);
            });
        }
    }

    abstract class BaseEntity
    {
        public int Id { get; set; }
    }

    class DerivedEntity : BaseEntity
    {
        public OwnedEntity OwnedEntity { get; set; } = new OwnedEntity();
    }

    class DerivedEntity2 : BaseEntity
    {
        public OwnedEntity2 OwnedEntity2 { get; set; } = new OwnedEntity2();
    }

    class OwnedEntity
    {
        public int IntValue { get; set; }
    }

    class OwnedEntity2
    {
        public string StringValue { get; set; }
    }
}
```

### Further technical details
EF Core version: 2.1.3
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10 Pro 1803
IDE: Visual Studio 2017 15.8.5

**Comment by gojanpaolo** [comment]
@uso800 Inheritance with owned types is one of the current shortcomings as [stated in the docs](https://docs.microsoft.com/en-us/ef/core/modeling/owned-entities#current-shortcomings).
> ### Current shortcomings
> - Inheritance hierarchies that include owned entity types are not supported

Also, might be related to: https://github.com/aspnet/EntityFrameworkCore/issues/9630

**Comment by maumar** [comment]
Problem here is that we try to eager load owned entities before we run QM optimizations. This results in `EagerLoadingExpressionVisitor` executing on a following, unoptimized QM:

```
(from BaseEntity <generated>_1 in DbSet<BaseEntity>
select [<generated>_1]).OfType<DerivedEntity>()
```

and tries to load all owned entities defined on all derived types.

Query optimized converts the QM to:

```
from DerivedEntity <generated>_1 in DbSet<DerivedEntity>
select [<generated>_1]
```

for which `OwnedEntity2` is not accessible.

We should be running `EagerLoadingExpressionVisitor` inside `QueryOptimizer`, not before.

**Comment by maumar** [comment]
fixed in 44184d2419dd12a0af9c9d6a9e6c0d2821aeaaac
